File: analysis/matlab/templates/example_analysis.m
%% ETRA 解析テンプレート — MATLAB 基本解析
% =============================================
% 使用方法:
%   1. このファイルを personal/<username>/ にコピー
%   2. BUCKET / INPUT_KEY / OUTPUT_PREFIX を変更
%   3. Amazon DCV 経由で EC2 上の MATLAB で実行
%
% 前提: MATLAB R2024b, AWS SDK for MATLAB (matlab-aws-sdk) または AWS CLI 経由

%% 設定
BUCKET        = 'etra-dev-workspace';
INPUT_KEY     = 'personal/<your-user-id>/input/data.csv';
OUTPUT_PREFIX = 'personal/<your-user-id>/results/';

%% S3 からデータ読み込み
% AWS CLI 経由で一時ファイルにダウンロード
tmp_input = tempname;
cmd = sprintf('aws s3 cp s3://%s/%s %s', BUCKET, INPUT_KEY, tmp_input);
status = system(cmd);
if status ~= 0
    error('S3 からのダウンロードに失敗しました: %s', INPUT_KEY);
end

data = readtable(tmp_input);
delete(tmp_input);
fprintf('データ読み込み完了: %d 行 x %d 列\n', height(data), width(data));

%% 基本統計
fprintf('\n=== 基本統計 ===\n');
disp(summary(data));

%% 数値列の解析
num_vars = data.Properties.VariableNames( ...
    varfun(@isnumeric, data, 'OutputFormat', 'uniform'));

if isempty(num_vars)
    warning('数値列が見つかりません');
    return;
end

x = data{:, num_vars{1}};
fprintf('\n対象列: %s\n', num_vars{1});
fprintf('  平均: %.4f\n', mean(x, 'omitnan'));
fprintf('  標準偏差: %.4f\n', std(x, 'omitnan'));
fprintf('  最大: %.4f\n', max(x));
fprintf('  最小: %.4f\n', min(x));

%% フィルタリング（バターワースフィルタ）
fs = 100;                          % サンプリング周波数 [Hz]
fc = 10;                           % カットオフ周波数 [Hz]
[b, a] = butter(4, fc/(fs/2));     % 4次バターワースローパス
x_filtered = filtfilt(b, a, x);

%% プロット
fig = figure('Visible', 'off', 'Position', [0 0 1000 600]);

subplot(2, 1, 1);
t = (0:length(x)-1) / fs;
plot(t, x, 'Color', [0.5 0.5 0.5], 'DisplayName', 'raw'); hold on;
plot(t, x_filtered, 'b', 'DisplayName', 'filtered'); hold off;
legend; xlabel('Time [s]'); ylabel('Amplitude');
title(sprintf('Signal: %s', num_vars{1}));

subplot(2, 1, 2);
[pxx, f] = pwelch(x_filtered, [], [], [], fs);
semilogy(f, pxx);
xlabel('Frequency [Hz]'); ylabel('PSD');
title('Power Spectral Density');

%% 結果を S3 に保存
% 図を PNG として保存
tmp_fig = [tempname '.png'];
exportgraphics(fig, tmp_fig, 'Resolution', 150);
close(fig);

fig_key = [OUTPUT_PREFIX 'signal_plot.png'];
cmd = sprintf('aws s3 cp %s s3://%s/%s --content-type image/png', ...
    tmp_fig, BUCKET, fig_key);
status = system(cmd);
delete(tmp_fig);
if status == 0
    fprintf('図を保存: s3://%s/%s\n', BUCKET, fig_key);
else
    warning('図の保存に失敗しました');
end

% 統計テーブルを CSV として保存
stats_table = table( ...
    {'mean'; 'std'; 'max'; 'min'}, ...
    [mean(x,'omitnan'); std(x,'omitnan'); max(x); min(x)], ...
    [mean(x_filtered,'omitnan'); std(x_filtered,'omitnan'); max(x_filtered); min(x_filtered)], ...
    'VariableNames', {'metric', 'raw', 'filtered'});

tmp_csv = [tempname '.csv'];
writetable(stats_table, tmp_csv);

stats_key = [OUTPUT_PREFIX 'stats.csv'];
cmd = sprintf('aws s3 cp %s s3://%s/%s --content-type text/csv', ...
    tmp_csv, BUCKET, stats_key);
status = system(cmd);
delete(tmp_csv);
if status == 0
    fprintf('統計を保存: s3://%s/%s\n', BUCKET, stats_key);
end

fprintf('\n解析完了\n');

File: analysis/python/templates/example_analysis.py
"""
ETRA 解析テンプレート — Python 基本解析
=========================================
使用方法:
  1. このファイルを personal/<username>/ にコピー
  2. INPUT_PREFIX / OUTPUT_PREFIX を自分のパスに変更
  3. JupyterLab または EC2 上で実行

依存: numpy, scipy, matplotlib, pandas, boto3
"""

import io
import os

import boto3
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy import signal

matplotlib.use("Agg")  # GUI なし環境向け

# ─── 設定 ───────────────────────────────────────────────
BUCKET = os.environ.get("WORKSPACE_BUCKET", "etra-dev-workspace")
INPUT_PREFIX = "personal/<your-user-id>/input/"
OUTPUT_PREFIX = "personal/<your-user-id>/results/"

s3 = boto3.client("s3")


# ─── S3 ヘルパー ─────────────────────────────────────────
def read_csv_from_s3(key: str) -> pd.DataFrame:
    """S3 から CSV を DataFrame として読み込む"""
    obj = s3.get_object(Bucket=BUCKET, Key=key)
    return pd.read_csv(io.BytesIO(obj["Body"].read()))


def save_figure_to_s3(fig: plt.Figure, key: str) -> None:
    """matplotlib Figure を PNG として S3 に保存"""
    buf = io.BytesIO()
    fig.savefig(buf, format="png", dpi=150, bbox_inches="tight")
    buf.seek(0)
    s3.put_object(Bucket=BUCKET, Key=key, Body=buf, ContentType="image/png")
    print(f"保存: s3://{BUCKET}/{key}")


def save_csv_to_s3(df: pd.DataFrame, key: str) -> None:
    """DataFrame を CSV として S3 に保存"""
    buf = io.BytesIO()
    df.to_csv(buf, index=False)
    buf.seek(0)
    s3.put_object(Bucket=BUCKET, Key=key, Body=buf, ContentType="text/csv")
    print(f"保存: s3://{BUCKET}/{key}")


# ─── 解析処理 ─────────────────────────────────────────────
def run_analysis() -> None:
    # ── サンプルデータ生成（実際は read_csv_from_s3 で読み込む）──
    t = np.linspace(0, 10, 1000)
    signal_data = np.sin(2 * np.pi * t) + 0.3 * np.random.randn(len(t))

    # ── フィルタリング ─────────────────────────────────────
    b, a = signal.butter(4, 0.1)
    filtered = signal.filtfilt(b, a, signal_data)

    # ── 統計量 ────────────────────────────────────────────
    stats = pd.DataFrame({
        "metric": ["mean", "std", "max", "min"],
        "raw": [signal_data.mean(), signal_data.std(), signal_data.max(), signal_data.min()],
        "filtered": [filtered.mean(), filtered.std(), filtered.max(), filtered.min()],
    })
    print(stats.to_string(index=False))

    # ── プロット ──────────────────────────────────────────
    fig, axes = plt.subplots(2, 1, figsize=(10, 6), sharex=True)
    axes[0].plot(t, signal_data, alpha=0.6, label="raw")
    axes[0].plot(t, filtered, label="filtered")
    axes[0].legend()
    axes[0].set_title("Signal")

    freqs, psd = signal.welch(filtered, fs=100)
    axes[1].semilogy(freqs, psd)
    axes[1].set_xlabel("Frequency [Hz]")
    axes[1].set_title("PSD")

    fig.tight_layout()

    # ── S3 に保存 ─────────────────────────────────────────
    save_figure_to_s3(fig, f"{OUTPUT_PREFIX}signal_plot.png")
    save_csv_to_s3(stats, f"{OUTPUT_PREFIX}stats.csv")
    plt.close(fig)


if __name__ == "__main__":
    run_analysis()
